spec/compliance_data_spec.rb

```diff
@@ -0,0 +1,162 @@
+require 'spec_helper'
+require 'stub_classes'
+
+describe ComplianceData do
+  describe 'initialize' do
+    it 'returns the right thing' do
+      data = ComplianceData.new StubClasses::StubSettings
+      expect(data).to be_an_instance_of ComplianceData
+    end
+
+    it 'raises an argument error if settings are not correct' do
+      expect { ComplianceData.new nil }.to raise_error 'ArgumentError'
+    end
+  end
+
+  describe 'base_name' do
+    @data = nil
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      base = @data.base_name 'path/to/file.json'
+      expect(base).to eq 'file'
+    end
+
+    it 'handles a missing extension' do
+      base = @data.base_name 'path/to/file'
+      expect(base).to eq 'file'
+    end
+
+    it 'handles a base name' do
+      base = @data.base_name 'file'
+      expect(base).to eq 'file'
+    end
+
+    it 'returns empty string if passed nil' do
+      base = @data.base_name nil
+      expect(base).to eq ''
+    end
+  end
+
+  describe 'full_name' do
+    @data = nil
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      full = @data.full_name 'file'
+      expect(full).to eq 'results/file.json'
+    end
+
+    it 'returns an empty string if passed nil' do
+      full = @data.full_name nil
+      expect(full).to eq ''
+    end
+  end
+
+  describe 'key_list' do
+    @data = nil
+
+    class StubObjectSummary
+      attr_reader :key
+      def initialize(key_value)
+        @key = key_value
+      end
+    end
+
+    before :each do
+      @data = ComplianceData.new StubClasses::StubSettings
+    end
+
+    it 'does the right thing' do
+      allow_any_instance_of(Aws::S3::Bucket).to receive(:objects)
+        .and_return(
+          [
+            StubObjectSummary.new(
+              "#{StubClasses::StubSettings.results_folder}/abc"),
+            StubObjectSummary.new(
+              "#{StubClasses::StubSettings.results_folder}/bcd")
+          ])
```

Comment:
I'd prefer to wrap the S3 bucket so the user doesn't need to handle the instantiation, and I'm hesitant to change that model just to make the class easier to test. I'd like to make it follow standard ruby practices though? Which way would be more ruby-like?